### PR TITLE
Fix report_reader to have several time the same index (node_id, compa…

### PR DESCRIPTION
**Pull request by alkino**
…rtment_id)

### CODE REVIEW

File: src/report_reader.cpp [hunk 1]
```diff
@@ -371,7 +371,7 @@ DataFrame<T> ReportReader<T>::Population::get(const Selection& selection,
             for (auto& datum : data) {
                 data_by_node.push_back(datum[i]);
             }
-            data_frame.data.insert(make_value<T>(node_id, element_ids[i], std::move(data_by_node)));
+            data_frame.data.push_back(make_value<T>(node_id, element_ids[i], std::move(data_by_node)));
```
Comment (cattabiani) [inline src/report_reader.cpp:374]: maybe emplace_back could be better

Comment (alkino) [reply src/report_reader.cpp]: As `make_value` return a R-value (prvalue to be correct), we don't care.

File: src/report_reader.cpp [hunk 2]
```diff
@@ -359,19 +355,21 @@ DataFrame<T> ReportReader<T>::Population::get(const Selection& selection,
             .select({index_start, it->second.first},
                     {index_stop - index_start + 1, it->second.second - it->second.first})
             .read(data);
+        int index = 0;
+        for (auto& datum : data) {
+            for (auto d : datum) {
+                data_frame.data[index].push_back(d);
+            }
```
Comment (tomdele) [inline src/report_reader.cpp:362]: I think changing the names would be beneficial for readability. 
I am a bit lost with the indexing order... 

datum is column or line from the initial 2D array ?
index is ids or times ?

Comment (tomdele) [reply src/report_reader.cpp:362]: You can also remove the second auto --> float. This makes clear why you don `&`

File: src/report_reader.cpp [hunk 3]
```diff
@@ -359,19 +355,21 @@ DataFrame<T> ReportReader<T>::Population::get(const Selection& selection,
             .select({index_start, it->second.first},
                     {index_stop - index_start + 1, it->second.second - it->second.first})
             .read(data);
+        int timer_index = 0;
+        for (const std::vector<float>& datum : data) {
+            for (float d : datum) {
+                data_frame.data[timer_index].push_back(d);
+            }
+            ++timer_index;
+        }
```
Comment (tomdele) [inline src/report_reader.cpp:364]: Sorry I am lost again, in python I have : 
data_frame.data[x][y] with x == nodes and y == times

You have the opposite here. Is this due to the C vs F representation in bindings ?

Comment (tomdele) [reply src/report_reader.cpp:364]: or  is  it `ids_index` instead ?

File: include/bbp/sonata/report_reader.h [hunk 4]
```diff
@@ -18,9 +18,10 @@ namespace sonata {
 // KeyType will be NodeID for somas report and pair<NodeID, uint32_t> for elements report
 template <typename KeyType>
 struct SONATA_API DataFrame {
-    using DataType = std::map<KeyType, std::vector<float>>;
-    std::vector<double> index;
-    DataType data;
+    using DataType = std::vector<KeyType>;
+    std::vector<double> times;
+    DataType ids;
+    std::vector<std::vector<float>> data;
```
Comment (tomdele) [inline include/bbp/sonata/report_reader.h:24]: once we are sure about the first and second indices for data we can add a comment here:
// data[ids][times] 
or 
// data[times][ids]

File: src/report_reader.cpp [hunk 5]
```diff
@@ -359,19 +355,21 @@ DataFrame<T> ReportReader<T>::Population::get(const Selection& selection,
             .select({index_start, it->second.first},
                     {index_stop - index_start + 1, it->second.second - it->second.first})
             .read(data);
+        int index = 0;
+        for (auto& datum : data) {
```
Comment (jorblancoa) [inline src/report_reader.cpp]:
```suggestion
        for (const auto& datum : data) {
```

**Comment by jdcourcol** [comment]
@alkino can this move forward ?

File: src/report_reader.cpp (hunk 3 above)
Comment (alkino) [reply src/report_reader.cpp:364]: I have no idea, it's  clearly difficult to tell, too many people read things as they want in the middle ^^

File: src/report_reader.cpp (hunk 5 above)
Comment (alkino) [reply src/report_reader.cpp]: This is a float now

File: src/report_reader.cpp (hunk 3 above)
Comment (alkino) [reply src/report_reader.cpp:364]: In fact if we look at the c++ tests it's clearly [times][ids]

**Comment by jorblancoa** [comment]
LGTM!

**Comment by tomdele** [comment]
I think we still lack quite a lot of testing here. 

There is no tests for the ElementReader ? 
You do not test any data values in the python part also. 

These need to be tested for sure.

File: src/report_reader.cpp (hunk 3 above)
Comment (tomdele) [reply src/report_reader.cpp:364]: Yes I checked again I messed up this is indeed [times][ids]

**Comment by alkino** [comment]
test are in python for ElementReader, test in cpp are more here to help me develop.
What do you mean by data values?

**Comment by tomdele** [comment]
I think you should test the C++ side intensively also. I guess some people (from viz or elsewhere) would prefer to use the C++ version. Also testing both with the same expected values can detect binding problems.

In the python part, you are testing some length from the data but never the outputted data themselves. There is nothing that guaranty you retrieve the correct data.

**Comment by tomdele** [comment]
Also did you drop the idea of numpy array for the data ?

**Comment by tomdele** [comment]
Also edge cases are missing on the python side : 
- empty list 
- numpy array 
- empty tuple 
- swapping ids : [1, 2] vs [2, 1]
- the out of range times
- the minus -1 ids
etc ... 

Also I found this using elements.h5 : 
```
rep.get(node_ids=[99999]).data                                                                                                                                                                                              
[[], [],  [],  [],  [],  [],  [],  [],  [],  [],  [],  [],  [],  [],  [],  [],  [],  [],  [], []]

rep.get(node_ids=[999999]).times                                                                                                                                                       
[0.0, 0.2, 0.4, 0.6000000000000001, 0.8, 1.0, 1.2, 1.4, 1.5999999999999999, 1.7999999999999998, 1.9999999999999998, 2.1999999999999997, 2.4, 2.6, 2.8000000000000003, 3.0000000000000004, 3.2000000000000006, 3.400000000000001, 3.600000000000001, 3.800000000000001]

rep.get(node_ids=[999999]).ids
[]
```
I am fine with this behavior but is it the one you intended ?

**Comment by mgeplf** [comment]
Two things:

1) What is the functional testing strategy for this?  Should it have beeen able to catch the problem solved by this PR?
2) In the future, could we give the opportunity to the other people who had participated in the  review to give their approval before merge - that way they can check if their satisfied as well.

**Comment by alkino** [comment]
I'm on it.

**Comment by alkino** [comment]
>     1. What is the functional testing strategy for this?  Should it have beeen able to catch the problem solved by this PR?

The problem fixed by this PR is not a bug, this is a new way to see data, so I fix test, because we misunderstood things. So the test check what we things as the good way to see things.

**Pull request merged**